stag-library/src/test/java/com/vimeo/stag/KnownTypeAdaptersTestKt.kt

```diff
@@ -268,4 +300,81 @@ class KnownTypeAdaptersTestKt {
         assertThat(KnownTypeAdapters.JSON_PRIMITIVE.fromJson(json)).isEqualTo(value)
     }
 
+    data class KeyTest(val string: String)
+
+    @Test
+    fun `MapTypeAdapter serializes and deserializes object keys correctly`() {
+        val mapTypeAdapter = KnownTypeAdapters.MapTypeAdapter<KeyTest, String, Map<KeyTest, String>>(
+                object : TypeAdapter<KeyTest>() {
+                    override fun write(out: JsonWriter, value: KeyTest?) {
+                        out.beginObject()
+                        out.name("string")
+                        out.value(value?.string)
+                        out.endObject()
+                    }
+
+                    override fun read(input: JsonReader): KeyTest {
+                        input.beginObject()
+                        input.nextName()
+                        val string = input.nextString()
+                        input.endObject()
+                        return KeyTest(string ?: throw Exception("Null string"))
+                    }
+
+                },
+                KnownTypeAdapters.STRING_NULL_SAFE_TYPE_ADAPTER,
+                ObjectConstructor<Map<KeyTest, String>> { HashMap() }
+        )
+
+        val testMap = mapOf(
+                KeyTest("a") to "b",
+                KeyTest("b") to "c"
+        )
+
+        val json = mapTypeAdapter.toJson(testMap)
+        assertThat(mapTypeAdapter.fromJson(json)).isEqualTo(testMap)
```

Comment:
I didn't dig into this too much, but on the surface it looks like you are only testing the `mapTypeAdapter` that you create in the test (with overridden `read` and `write` methods). Is there any part of `MapTypeAdapter` that is actually being tested here? If so might be worth adding an inline comment somewhere explaining what part of `MapTypeAdapter` is being tested. Same applies to the next unit test.

Reply:
It's a little difficult to tell because the instantiation is verbose, but the `read` and `write` methods are from the `TypeAdapter` that is being _injected_ into the `MapTypeAdapter`. So these tests are verifying that the adapter properly delegates to the injected `TypeAdapter` depending on the type used. I broke out the adapter into its own variable to make it more clear.